Generics & ArrayList

MarcelGarus opened the pull request.

MarcelGarus committed:
Add basic constructor support

MarcelGarus committed:
Fix minor errors

MarcelGarus committed:
Implement ArrayList

MarcelGarus committed:
Merge remote-tracking branch 'origin/feature/comparisons' into feature/collections

MarcelGarus committed:
Lay groundwork for baking type arguments

MarcelGarus committed:
Merge remote-tracking branch 'origin/feature/comparisons' into feature/collections

MarcelGarus committed:
Add missing type parameter

MarcelGarus committed:
Makea getAllDependencies also return the core package

MarcelGarus committed:
Comment out use of non-existent module

MarcelGarus committed:
Add typeParameters to function call and constructor call

MarcelGarus committed:
Add bakeGenerics method to CandyType

MarcelGarus committed:
Implement loweConstructorCall

MarcelGarus committed:
Minor fixes

MarcelGarus committed:
Add id to ClassDeclaration

MarcelGarus committed:
Add ConstructorCallExpression

MarcelGarus committed:
Add ugly workaround for generics

MarcelGarus committed:
Merge branch 'master' into feature/collections

MarcelGarus committed:
Merge remote-tracking branch 'origin/feature/arithmetic-operators' into feature/collections

MarcelGarus committed:
Merge branch 'master' into feature/collections

MarcelGarus committed:
Improve ParameterCandyType's toString

MarcelGarus committed:
Make ugly generic workaround a bit more permissive

MarcelGarus committed:
Debug core

MarcelGarus committed:
Implement constructor calls

MarcelGarus committed:
Fix ambiguity errors

MarcelGarus committed:
Fix ambiguity errors even more

MarcelGarus committed:
Make ArrayList's length a function

MarcelGarus committed:
Gracefully handle baking function types without a receiver type

MarcelGarus committed:
Improve ambiguity error

MarcelGarus committed:
Merge branch 'master' into feature/collections

MarcelGarus committed:
Don't punish using the class_ identifier

MarcelGarus committed:
Bake types when referencing properties in impls

MarcelGarus committed:
Add unwrap method to Option

MarcelGarus committed:
Remove unprefixed hir import

MarcelGarus committed:
Properly bake function types

MarcelGarus committed:
Fix some of the array code

MarcelGarus committed:
Rename Array.fill to Array.filled

MarcelGarus committed:
Make UserType never assignable to ParameterType

MarcelGarus committed:
Properly bake UserTypes in isAssignableTo

MarcelGarus committed:
Fix comment

MarcelGarus committed:
Lower Array.filled

MarcelGarus committed:
Merge branch 'master' into feature/collections

MarcelGarus committed:
Fix merged code

MarcelGarus committed:
Remove debug outputs

MarcelGarus committed:
Add expression and code as getters on Method

MarcelGarus committed:
Call List.filled function correctly

MarcelGarus committed:
Support type parameters in function call

MarcelGarus committed:
Support type parameters in constructor call

MarcelGarus committed:
Remove faulty explicit type annotations in final/var assignments

MarcelGarus committed:
Support identifiers that start with "in" (i.e., index)

MarcelGarus committed:
In Dart, represent Unit as void rather than Null

MarcelGarus committed:
Remove legacy builtins for of1, of2, etc.

MarcelGarus committed:
Fix ArrayList

MarcelGarus committed:
Make code more idiomatic

MarcelGarus committed:
Implement isEmpty and isNotEmpty

MarcelGarus committed:
Rename MutableList's add to append to avoid naming conflict with the add operator

MarcelGarus committed:
Fix assignment lowering to Dart

JonasWanke commented on packages/candy/src/collections/list/array.candy:
```diff
@@ -1,26 +1,137 @@
 use ....primitives
 use ...list
 
+builtin class Array<Blub> {
```
The type parameter can be renamed back to `Item`

JonasWanke commented on packages/candy/src/collections/list/array.candy:
```diff
@@ -1,26 +1,137 @@
+  //   list.add(fourth)
+  //   list.add(fifth)
+  //   list
+  // }
```
Why are these commented out?

JonasWanke commented on packages/candy/src/collections/list/array.candy:
```diff
@@ -1,26 +1,137 @@
+  //   list
+  // }
+
+  static fun create<Item>(capacity: Int /* Uint */): ArrayList<Item> {
```
```suggestion
  static fun create<Item>(capacity: Int /* UInt */): ArrayList<Item> {
```

JonasWanke commented on packages/candy/src/collections/list/array.candy:
```diff
@@ -1,26 +1,137 @@
+  // TODO(marcelgarus): Getterize!
+  fun capacity(): Int /* UInt */ {
+    items.length
+  }
```
```suggestion
  fun capacity(): Int /* UInt */ { items.length }
```

JonasWanke commented on packages/candy/src/collections/list/array.candy:
```diff
@@ -1,26 +1,137 @@
-// }
+impl<Item> ArrayList<Item>: List<Item> {
+  fun length(): Int { length_ }
+  fun isEmpty(): Bool { length_ == 0 }
```
Can we move this to `List`?

JonasWanke commented on packages/candy/src/collections/list/array.candy:
```diff
@@ -1,26 +1,137 @@
 
-builtin class Array<Item> {
-  let length: Int
+  fun get(index: Int /* UInt */): Item {
```
Should return an `Option<Item>`

JonasWanke commented on packages/candy/src/collections/list/module.candy:
```diff
@@ -1,27 +1,35 @@
+  }
+  fun isNotEmpty(): Bool {
+    !isEmpty()
+  }
```
```suggestion
  fun isEmpty(): Bool { length() == 0 }
  fun isNotEmpty(): Bool { !isEmpty() }
```

JonasWanke commented on packages/candy/src/collections/list/module.candy:
```diff
@@ -1,27 +1,35 @@
+  // builtin static fun of2<Item>(first: Item, second: Item): ArrayList<Item>
+  // builtin static fun of3<Item>(first: Item, second: Item, third: Item): ArrayList<Item>
+  // builtin static fun of4<Item>(first: Item, second: Item, third: Item, fourth: Item): ArrayList<Item>
+  // builtin static fun of5<Item>(first: Item, second: Item, third: Item, fourth: Item, fifth: Item): ArrayList<Item>
```
Why are these commented out?

JonasWanke commented on packages/candy/src/collections/list/module.candy:
```diff
@@ -1,27 +1,35 @@
-  fun indexOf(item: Item): Option<Int> /* Option<UInt> */
-  fun lastIndexOf(item: Item): Option<Int> /* Option<UInt> */
-  // TODO(JonasWanke): containsAll?
+  fun get(index: Int /* UInt */): Item
```
Should return an `Option<Item>`

JonasWanke commented on packages/candy/src/primitives/option.candy:
```diff
@@ -1,10 +1,20 @@
+impl<Value> Some<Value>: Option<Value> {
+  fun unwrap(): Value {
+    value
+  }
```
```suggestion
  fun unwrap(): Value { value }
```

JonasWanke commented on packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart:
```diff
@@ -1,3 +1,5 @@
+import 'dart:io';
+
```
```suggestion
```

JonasWanke commented on packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart:
```diff
@@ -1665,8 +1735,15 @@ extension on Context {
+
       throw CompilerError.unsupportedFeature(
-        'Callable expressions are not yet supported.',
+        'Callable expressions are not yet supported (target $target).',
```
```suggestion
        'Callable expressions are not yet supported (target: $target).',
```

JonasWanke commented on packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart:
```diff
@@ -1785,7 +1903,111 @@ extension on Context {
+  ) {
+    assert(target != null);
+    assert(target.identifier is hir.ReflectionIdentifier,
+        'target.identifier is not a ReflectionIdentifier: ${target.identifier}');
```
```suggestion
        'target.identifier is not a `ReflectionIdentifier`: ${target.identifier}');
```

JonasWanke commented on packages/compiler/lib/src/compilation/hir/expressions.dart:
```diff
@@ -1,3 +1,4 @@
+import 'package:compiler/compiler.dart';
```
This should not be imported here.

JonasWanke commented on packages/compiler_dart/lib/src/body.dart:
```diff
@@ -156,6 +157,39 @@ class DartExpressionVisitor extends ExpressionVisitor<List<dart.Code>> {
+          final t = dart.refer('T');
+          final name = _name(node.id);
+          final function = dart.Method((b) => b
+            ..name = name
```
```suggestion
          final function = dart.Method((b) => b
            ..name = _name(node.id)
```

MarcelGarus committed:
Represent trait's upper bounds in Dart

MarcelGarus committed:
Fix assignment of Unit types in Dart

MarcelGarus committed:
Add todo

MarcelGarus committed:
Move some ArrayList functionality to Iterable

MarcelGarus committed:
Merge branch 'master' into feature/collections

MarcelGarus committed:
Fix Dart SDK constraints

MarcelGarus committed:
Correctly navigate to functions defined in impls or traits

Including properly baking the types

MarcelGarus committed:
Minor fixes

JonasWanke commented on packages/candy/src/collections/iterable.candy:
```diff
@@ -1,14 +1,85 @@
+  // TODO(marcelgarus): expand
+  // TODO(marcelgarus): firstWhere
+  // TODO(marcelgarus): fold
+  // TODO(marcelgarus): followedBy
```
This could be done with the add operator

JonasWanke commented on packages/candy/src/collections/iterable.candy:
```diff
@@ -1,14 +1,85 @@
+  // TODO(marcelgarus): every
+  // TODO(marcelgarus): expand
+  // TODO(marcelgarus): firstWhere
+  // TODO(marcelgarus): fold
```
Maybe separate `foldLeft`/`foldRight`? (Though `foldRight` can be accomplished by first doing a `reverse`, which we also want)

JonasWanke commented on packages/candy/src/collections/iterable.candy:
```diff
@@ -1,14 +1,85 @@
+  // TODO(marcelgarus): lastWhere
+  // TODO(marcelgarus): map
+  // TODO(marcelgarus): enumerate
+  // TODO(marcelgarus): reduce
```
See `fold`

MarcelGarus committed:
Rename Array's type parameter from Blub to Item

MarcelGarus committed:
Minor style improvements

JonasWanke commented on packages/compiler_dart/lib/src/body.dart:
```diff
@@ -518,15 +520,19 @@ class DartExpressionVisitor extends ExpressionVisitor<List<dart.Code>> {
   @override
   List<dart.Code> visitWhileExpression(WhileExpression node) => [
-        dart.Code('void ${_name(node.id)};'),
+        dart.literalNull
+            .assignVar(_name(node.id), compileType(context, node.type))
+            .statement,
```
```suggestion
        _save(node, dart.literalNull, isMutable: true),
```

JonasWanke commented on packages/compiler_dart/lib/src/body.dart:
```diff
@@ -559,48 +565,37 @@ class DartExpressionVisitor extends ExpressionVisitor<List<dart.Code>> {
+        } else if (getPropertyDeclarationHir(context, property.id).isStatic) {
+          return compileTypeName(context, parent).property(name);
+        } else {
+          assert(property.receiver != null);
+          code.addAll(property.receiver.accept(this));
+          return dart.refer(name);
```
This only works when the receiver is `this`, but not for, e.g., `foo.bar = 1`

MarcelGarus committed:
Lift isEmpty implementations up

Iterable now defined isEmpty and isNotEmpty directly.
List now defines isEmpty using its length.
ArrayList doesn't define isEmpty anymore.

MarcelGarus committed:
Minor cleanups

MarcelGarus committed:
Add of1 to of5 to ArrayList

MarcelGarus committed:
Add very basic constraint solving in a few places

MarcelGarus committed:
Add of1 to of5 methods to List

MarcelGarus committed:
Update a few of the Iterable methods
